### ISSUE TITLE
[API Proposal]: `ValueTextEquals` supporting `JsonEncodedText` as a parameter

### ISSUE THREAD

**Issue by YohDeadfall** [dotnet/runtime#86591]
### Background and motivation

Currently there's an asymmetry between `Utf8JsonReader` and `Utf8JsonWriter`. Only the last one has API which accept `JsonEncodedText` while the first one doesn't. In the result one has to access the `EncodedUtf8Bytes` property when they write a custom converter with cached property names.

### API Proposal

```diff
namespace System.Collections.Generic;

public struct Utf8JsonReader
{
    public bool ValueTextEquals(JsonEncodedText utf8Text)
        => ValueTextEquals(utf8Text.EncodedUtf8Bytes);
}
```

### API Usage

```csharp
private static readonly JsonEncodedText PropertyName = JsonEncodedText.Encode("property");

writer.WriteString(PropertyName, "We already have that");
// ...
// // Now let's do that
reader.ValueTextEquals(PropertyName);
// instead of
reader.ValueTextEquals(PropertyName.EncodedUtf8Bytes);
```

**Comment by krwq** [comment]
Thank you for your proposal. However, we believe that the proposed ValueTextEquals function should not accept JsonEncodedText as a parameter. The purpose of JsonEncodedText is to efficiently handle pre-encoded JSON key strings, and not to provide a way to compare JSON values. The proposed usage seems to be mixing these two different concepts. Therefore, we do not plan on implementing this suggestion as it diverges from the intended purpose of these APIs​

**Comment by YohDeadfall** [comment]
The idea of that proposal isn't different from efficiently handling JSON keys, but it addresses the reader side.

As an example, Elastic heavily relies on custom JSON converters. Some of them look for properties, read their values and then construct objects. Currently it uses `GetValue` on property name tokens which is quite inefficient. I improved it a bit, so instead of decoding and allocating `string` the client just checks does the current value matches some property name or not via `ValueTextEquals` as in the example above.

With this proposal it would be a bit easier and maybe even more obvious that instead of strings `JsonEncodedText` can be used for property name checks.

[Read](https://github.com/elastic/elasticsearch-net/blob/7db556afac969d7d3d5d5dc2050adcea1a0fb633/src/Elastic.Clients.Elasticsearch/Core/Infer/Field/FieldConverter.cs#L54-L63) (could be without `EncodedUtf8Bytes` with that proposal):

```cs
                if (reader.ValueTextEquals(FieldProperty.EncodedUtf8Bytes))
                {
                    reader.Read();
                    field = reader.GetString();
                }
                else if (reader.ValueTextEquals(FormatProperty.EncodedUtf8Bytes))
                {
                    reader.Read();
                    format = reader.GetString();
                }
```

[Write](https://github.com/elastic/elasticsearch-net/blob/7db556afac969d7d3d5d5dc2050adcea1a0fb633/src/Elastic.Clients.Elasticsearch/Core/Infer/Field/FieldConverter.cs#L92-L93)
```cs
            writer.WriteString(FieldProperty, fieldName);
            writer.WriteString(FormatProperty, value.Format);
```

To summarize, supporting `JsonEncodedText` would improve experience of `Utf8JsonReader` usage, but that's a cosmetic change. Therefore, I won't insist on implementing it.

**Comment by krwq** [comment]
We understand that supporting JsonEncodedText could potentially simplify and improve the usage experience of Utf8JsonReader. However, our primary concern remains the same: JsonEncodedText was designed with the specific intent of handling pre-encoded JSON key strings for efficient writing, and not for reading or comparing JSON values.

As you pointed out, the proposed change would indeed be a cosmetic one. It would also blur the line between the distinct roles of JSON reading and writing, which is something we'd prefer to avoid. We believe it's important to maintain the clear separation between these functionalities to avoid potential misuse and confusion in the future.

Also it could be easily implemented as an extension method.
